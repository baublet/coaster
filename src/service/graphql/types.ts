import { ServiceType, ServiceDefaultProperties } from "../types";
import { Controller } from "../controller/types";
import { ModelFactory } from "model";

export type GraphQLPrimitiveTypes = {
  type: GraphQLPrimitiveType;
  nullable?: boolean;
};

export type GraphQLModelCollectionType = {
  type: GraphQLType.MODEL_COLLECTION;
  modelFactory: ModelFactory;
  nullable?: boolean;
};

export type GraphQLArrayOfType = {
  type: GraphQLType.ARRAY_OF;
  value: GraphQLReturnStructureNode;
  nullable?: boolean;
};

export type GraphQLObjectType = {
  type: GraphQLType.OBJECT;
  nodes: Record<string, GraphQLReturnStructureNode>;
  nullable?: boolean;
  name?: string;
  description?: string;
};

type GraphQLEnumFieldsMap = Record<
  string,
  | string
  | {
      deprecationReason?: string;
      description: string;
    }
>;

export type GraphQLEnumType = {
  type: GraphQLType.ENUM;
  nullable?: boolean;
  name?: string;
  description?: string;
  values: GraphQLEnumFieldsMap;
};

export enum GraphQLType {
  ARRAY_OF = "ARRAY_OF",
  BOOLEAN = "BOOLEAN",
  ENUM = "ENUM",
  FLOAT = "FLOAT",
  ID = "ID",
  INT = "INT",
  MODEL_COLLECTION = "MODEL_COLLECTION",
  MODEL = "MODEL",
  OBJECT = "OBJECT",
  SCALAR = "SCALAR",
  STRING = "STRING"
}

export type GraphQLPrimitiveType =
  | GraphQLType.SCALAR
  | GraphQLType.BOOLEAN
  | GraphQLType.STRING
  | GraphQLType.FLOAT
  | GraphQLType.ID
  | GraphQLType.INT;

export type GraphQLReturnStructureNode =
  | GraphQLPrimitiveTypes
  | GraphQLModelCollectionType
  | GraphQLArrayOfType
  | GraphQLObjectType
  | GraphQLEnumType;

// We omit models and model collections here. They don't make sense as arguments
export type GraphQLResolverArgument =
  | GraphQLPrimitiveTypes
  | GraphQLArrayOfType
  | GraphQLObjectType;

export type GraphQLResolverArguments = Record<string, GraphQLResolverArgument>;

export type ArgumentTypeFromArguments<
  T extends GraphQLResolverArguments
> = OptionalNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>> &
  RequiredNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>>;

export type GraphQLReturnStructureRootNode = GraphQLReturnStructureNode & {
  typename?: string;
};

export interface GraphQLQueryControllerConfiguration<
  ResolverArguments extends GraphQLResolverArguments = any,
  ReturnStructure extends GraphQLReturnStructureNode = any
> {
  description?: string;
  resolver: Controller<ResolverArguments, ReturnStructure>;
  resolverArguments?: GraphQLResolverArguments;
  resolutionType: GraphQLReturnStructureRootNode;
}

export type GraphQLQueryOrMutationNode = {
  [key: string]: GraphQLQueryControllerConfiguration;
};

export interface GraphQLServiceOptions {
  queries: GraphQLQueryOrMutationNode;
  mutations?: GraphQLQueryOrMutationNode;
}

export interface GraphQLServiceArguments extends ServiceDefaultProperties {
  type: ServiceType.GRAPHQL;
  options: GraphQLServiceOptions;
}

export type GraphQLPrimitiveEnumToType<
  T extends GraphQLPrimitiveType
> = T extends GraphQLType.STRING
  ? string
  : T extends GraphQLType.SCALAR
  ? any
  : T extends GraphQLType.FLOAT
  ? number
  : T extends GraphQLType.INT
  ? number
  : T extends GraphQLType.ID
  ? string
  : T extends GraphQLType.BOOLEAN
  ? boolean
  : never;

export type GraphQLObjectDeclarationNodes = Record<
  string,
  GraphQLReturnStructureNode
>;

export type GraphQLObjectDeclarationTypes<
  T extends GraphQLObjectDeclarationNodes
> = {
  [K in keyof T]: ReturnNodeToType<T[K]>;
};

export type GraphQLObjectDeclarationToType<
  T extends GraphQLObjectDeclarationNodes
> = RequiredNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>> &
  OptionalNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>>;

export type RequiredKeysInNodes<T extends GraphQLObjectDeclarationNodes> = {
  [K in keyof T]: T[K]["nullable"] extends false ? K : never;
}[keyof T];

export type OptionalNodes<
  T extends GraphQLObjectDeclarationNodes,
  K extends keyof T
> = Partial<Omit<T, K>>;

export type RequiredNodes<
  T extends GraphQLObjectDeclarationNodes,
  K extends keyof T
> = Pick<T, K>;

export type ReturnNodeToType<
  T extends GraphQLReturnStructureNode
> = T extends GraphQLPrimitiveTypes
  ? GraphQLPrimitiveEnumToType<T["type"]>
  : T extends GraphQLObjectType
  ? GraphQLObjectDeclarationToType<T["nodes"]>
  : T extends GraphQLArrayOfType
  ? ReturnNodeToType<T["value"]>[]
  : T extends GraphQLEnumType
  ? keyof T["values"]
  : never;

export type ReturnNodeToTypeOptionality<
  T extends any
> = T["nullable"] extends false ? undefined : T;

export type GraphQLTypedResolverDeclaration<
  ResolverArguments extends GraphQLResolverArguments,
  ReturnTypeArgument extends GraphQLReturnStructureNode
> = GraphQLQueryControllerConfiguration<
  ArgumentTypeFromArguments<ResolverArguments>,
  ReturnNodeToTypeOptionality<ReturnNodeToType<ReturnTypeArgument>>
>;
