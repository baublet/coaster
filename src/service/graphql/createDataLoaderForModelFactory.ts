import DataLoader from "dataloader";

import { PersistedModelFactory } from "persist";

export function createDataLoaderForModelFactory<
  T extends PersistedModelFactory
>(factory: T): DataLoader<string, ReturnType<T>> {
  return new DataLoader(
    (keys: string[]) => factory.find(keys) as Promise<(ReturnType<T> | null)[]>
  );
}
