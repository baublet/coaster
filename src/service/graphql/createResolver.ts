import {
  GraphQLResolverArguments,
  GraphQLTypedResolverDeclaration,
  GraphQLReturnStructureNode,
  ReturnNodeToType,
  ArgumentTypeFromArguments
} from "./types";
import { Controller } from "service/controller/types";

/**
 * Creates a GraphQL resolver from a specified resolver configuration. Enforces
 * strongly-typed resolvers.
 */
export function createResolver<
  ResolverArguments extends GraphQLResolverArguments,
  ReturnNode extends GraphQLReturnStructureNode
>({
  description,
  resolutionType,
  resolverArguments,
  resolver
}: {
  description?: string;
  resolverArguments?: ResolverArguments;
  resolutionType: ReturnNode;
  resolver: Controller<
    ArgumentTypeFromArguments<ResolverArguments>,
    ReturnNodeToType<ReturnNode>
  >;
}): GraphQLTypedResolverDeclaration<ResolverArguments, ReturnNode> {
  return {
    description,
    resolver,
    resolutionType,
    resolverArguments
  };
}
