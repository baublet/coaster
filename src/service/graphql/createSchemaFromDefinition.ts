import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLFloat,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLID,
  GraphQLFieldConfig,
  GraphQLList,
  GraphQLBoolean,
  GraphQLEnumType,
  GraphQLEnumValueConfigMap
} from "graphql";
import {
  GraphQLServiceArguments,
  GraphQLReturnStructureNode,
  GraphQLType,
  GraphQLObjectType as CoasterGraphQLObjectType,
  GraphQLEnumType as CoasterGraphQLEnumType
} from "./types";

function getPrimitiveFieldOptions(field: GraphQLReturnStructureNode) {
  switch (field.type) {
    case GraphQLType.FLOAT:
      return GraphQLFloat;
    case GraphQLType.STRING:
      return GraphQLString;
    case GraphQLType.INT:
      return GraphQLInt;
    case GraphQLType.ID:
      return GraphQLID;
    case GraphQLType.BOOLEAN:
      return GraphQLBoolean;
    default:
      throw new Error(`Unsupported GraphQL type: ${field.type}`);
  }
}

export function createSchemaFromDefinition<T extends GraphQLServiceArguments>(
  service: T
): GraphQLSchema {
  const objectTypes: Record<string, GraphQLObjectType> = {};
  // const scalars = {};
  const enumTypes = {};
  // const interfaces = {};

  const {
    queries: queryDefinitions = {}
    // mutations: mutationDefinitions = {}
  } = service.options;
  const queryKeys = Object.keys(queryDefinitions);
  // const mutationKeys = Object.keys(mutationDefinitions);

  /**
   * Takes a Coaster GraphQL enum type and converts it into a GraphQL enum
   * class object. Stores them on a global field to prevent duplicates.
   * @param object
   * @param name
   */
  function collateEnum(
    enumConfig: CoasterGraphQLEnumType,
    name?: string
  ): GraphQLEnumType {
    if (!name && !enumConfig.name) {
      throw new Error(
        "GraphQL enum types require a name. This error usually happens when we forget to give a name to root nodes. Otherwise, we infer names based on the object key."
      );
    }

    const typeName = enumConfig.name || name;
    if (enumTypes[typeName]) return enumTypes[typeName];

    const values: GraphQLEnumValueConfigMap = {};

    for (const [key, config] of Object.entries(enumConfig.values)) {
      let description: string;
      let deprecationReason: string;
      if (typeof config === "string") {
        description = config;
      } else {
        description = config.description;
        deprecationReason = config.deprecationReason;
      }
      values[key] = {
        description,
        deprecationReason
      };
    }

    return new GraphQLEnumType({
      name: typeName,
      values
    });
  }

  /**
   * Loops through the fields of a Coaster GraphQL object type definition,
   * collates all of the child types (recursively), and returns a fully-formed
   * GraphQLObjectType.
   * @param object
   * @param name
   */
  function collateObjectTypes(
    object: CoasterGraphQLObjectType,
    name?: string
  ): GraphQLObjectType {
    if (!name && !object.name) {
      throw new Error(
        "GraphQL object types require a name. This error usually happens when we forget to give a name to root nodes. Otherwise, we infer names based on the object key."
      );
    }

    const typeName = object.name || name;
    if (objectTypes[typeName]) return objectTypes[typeName];

    const fields = {};

    const fieldKeys = Object.keys(object.nodes);
    fieldKeys.forEach(key => {
      const fieldConfig: GraphQLFieldConfig<any, any> = {
        type: collateType(object.nodes[key], key)
        // TODO: allow field-level resolvers here
      };
      fields[key] = fieldConfig;
    });

    objectTypes[typeName] = new GraphQLObjectType({
      name: typeName,
      fields,
      description: object.description
    });

    return objectTypes[typeName];
  }

  /**
   * Takes an arbitrary GQL structure definition and returns a valid GraphQL
   * type object.
   * @param node
   * @param key
   */
  function collateType(node: GraphQLReturnStructureNode, key?: string) {
    let type;
    switch (node.type) {
      case GraphQLType.OBJECT:
        type = collateObjectTypes(node, key);
        break;
      case GraphQLType.ARRAY_OF:
        type = new GraphQLList(collateType(node.value));
        break;
      case GraphQLType.ENUM:
        type = collateEnum(node, key);
        break;
      default:
        type = getPrimitiveFieldOptions(node);
    }
    return node.nullable === false ? new GraphQLNonNull(type) : type;
  }

  // Step 1: Build the query fields
  const queryFields: Record<string, GraphQLFieldConfig<any, any>> = {};
  queryKeys.forEach(key => {
    const definition = queryDefinitions[key];
    const type = collateType(definition.resolutionType);
    queryFields[key] = {
      type,
      resolve: definition.resolver
    };
  });

  // Step 2: Build the mutation fields

  return new GraphQLSchema({
    types: Object.values(objectTypes),
    query: new GraphQLObjectType({
      name: "Query",
      fields: queryFields
    })
  });
}
