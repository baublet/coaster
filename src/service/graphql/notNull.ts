import { GraphQLReturnStructureNode } from "./types";

/**
 * Takes a Coaster GraphQL type and returns a copy of it, but as not null.
 * @param node
 */
export function notNull<T extends GraphQLReturnStructureNode>(
  node: T
): GraphQLReturnStructureNode {
  return Object.assign({}, node, { nullable: false });
}
