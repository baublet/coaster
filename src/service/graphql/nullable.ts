import { GraphQLReturnStructureNode } from "./types";

/**
 * Takes a Coaster GraphQL type and returns a copy of it, but as not null.
 * @param node
 */
export function nullable<T extends GraphQLReturnStructureNode>(node: T): T {
  return Object.assign({}, node, { nullable: true }) as T & {
    nullable: true;
  };
}
