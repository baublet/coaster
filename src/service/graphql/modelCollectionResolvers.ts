import { GraphQLInputObjectType, GraphQLInt } from "graphql";

export const CollectionInput = new GraphQLInputObjectType({
  name: "ModelCollectionInput",
  description: "Filtering, sorting, and pagination for model collections",
  fields: {
    limit: {
      type: GraphQLInt
    },
    offset: {
      type: GraphQLInt
    }
  }
});

export function rootCollectionResolver() {}
