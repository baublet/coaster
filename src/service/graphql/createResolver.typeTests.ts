import {
  GraphQLServiceArguments,
  GraphQLType,
  GraphQLReturnStructureNode
} from "./types";
import { ServiceType } from "service/types";
import { createResolver } from "./createResolver";

const queryAsAVariable = createResolver({
  description: "Resolver that is a variable",
  resolutionType: {
    type: GraphQLType.STRING
  },
  resolver: async () => {
    return "yo";
  }
});

const typeAsAVariable: GraphQLReturnStructureNode = {
  type: GraphQLType.OBJECT,
  nodes: {
    id: {
      type: GraphQLType.ID,
      nullable: false
    },
    name: {
      type: GraphQLType.STRING
    }
  }
};

export const _graphqlServiceArgumentsPrimitiveReturn: GraphQLServiceArguments = {
  name: "name",
  port: 82,
  type: ServiceType.GRAPHQL,
  options: {
    queries: {
      queryAsAVariable,
      test: createResolver({
        description: "description",
        resolverArguments: {
          id: {
            type: GraphQLType.STRING,
            nullable: false
          },
          count: {
            type: GraphQLType.INT
          },
          typeAsAVariable
        },
        resolutionType: {
          type: GraphQLType.INT
        },
        resolver: async ({ args }) => {
          /**
           * TODO: Make `return args.count` error.
           *
           * This condition should raise an error since count is an
           * optional property on arguments here...
           *
           * return args.count;
           */
          if (args.id) return 2;
          return 1;
        }
      })
    }
  }
};

export const _graphqlServiceArgumentsComplexReturns: GraphQLServiceArguments = {
  name: "name",
  port: 82,
  type: ServiceType.GRAPHQL,
  options: {
    queries: {
      test: createResolver({
        description: "description",
        resolverArguments: {
          id: {
            type: GraphQLType.STRING,
            nullable: false
          }
        },
        resolutionType: {
          type: GraphQLType.OBJECT,
          nodes: {
            counter: { type: GraphQLType.INT },
            text: { type: GraphQLType.STRING },
            scalar: { type: GraphQLType.SCALAR },
            boolean: { type: GraphQLType.BOOLEAN },
            list: {
              type: GraphQLType.ARRAY_OF,
              nullable: false,
              value: {
                type: GraphQLType.BOOLEAN
              }
            },
            nonNullableScalar: {
              type: GraphQLType.SCALAR,
              nullable: false
            },
            nonNullableFloat: {
              type: GraphQLType.FLOAT,
              nullable: false
            }
          }
        },
        resolver: async ({ args }) => {
          return {
            // text: 1, // Fails
            text: args.id,
            nonNullableScalar: "test",
            // nonNullableFloat: "test", // Fails
            nonNullableFloat: 1,
            // counter: {}, // Fails
            counter: 2,
            list: []
          };
        }
      })
    }
  }
};

export const _graphqlServiceArgumentsComplexReturns2: GraphQLServiceArguments = {
  name: "name",
  port: 82,
  type: ServiceType.GRAPHQL,
  options: {
    queries: {
      test: createResolver({
        description: "description",
        resolutionType: {
          type: GraphQLType.INT,
          nullable: false
        },
        resolverArguments: {
          id: {
            type: GraphQLType.STRING,
            nullable: false
          }
        },
        resolver: async () => {
          return ;
        }
      })
    }
  }
};
