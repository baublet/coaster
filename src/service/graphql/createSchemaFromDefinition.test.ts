import { printSchema } from "graphql";

import { ServiceType } from "service/types";
import { createSchemaFromDefinition } from "./createSchemaFromDefinition";
import { GraphQLType } from "./types";
import { createResolver } from "./createResolver";
import { createType } from "./createType";
import { notNull } from "./notNull";
import { User } from "testHelpers/User";

const TestObject = createType({
  type: GraphQLType.OBJECT,
  name: "TestCollection",
  nullable: false,
  description: "This is a test description",
  nodes: {
    totalCount: {
      type: GraphQLType.INT,
      nullable: false
    },
    nodes: {
      type: GraphQLType.INT
    }
  }
});

it("creates a schema", () => {
  expect(
    printSchema(
      createSchemaFromDefinition({
        name: "name",
        port: 82,
        type: ServiceType.GRAPHQL,
        options: {
          queries: {
            queryWithObjectReturn: createResolver({
              description: "description",
              resolutionType: notNull(TestObject),
              resolver: async () => {
                return null;
              }
            }),
            queryWithPrimitiveReturn: {
              resolutionType: {
                type: GraphQLType.BOOLEAN
              },
              resolver: async () => {}
            },
            queryWithListReturn: {
              resolutionType: {
                type: GraphQLType.ARRAY_OF,
                value: {
                  type: GraphQLType.OBJECT,
                  name: "TestObjectWithinAList",
                  nodes: {
                    testListOfObjects: {
                      type: GraphQLType.ID,
                      nullable: false
                    },
                    testModel: {
                      type: GraphQLType.MODEL,
                      modelFactory: User,
                      description: "Some user here"
                    },
                    enum: {
                      type: GraphQLType.ENUM,
                      name: "SomeTestEnum",
                      values: {
                        ENUM: "description",
                        ANOTHER: "another description"
                      }
                    }
                  }
                }
              },
              resolver: async () => {}
            }
          }
        }
      })
    )
  ).toMatchSnapshot();
});
