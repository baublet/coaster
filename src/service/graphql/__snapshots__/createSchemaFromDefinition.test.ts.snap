// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a schema 1`] = `
"type Query {
  queryWithObjectReturn: TestCollection!
  queryWithPrimitiveReturn: Boolean
  queryWithListReturn: [TestObjectWithinAList]
}

enum SomeTestEnum {
  \\"\\"\\"description\\"\\"\\"
  ENUM

  \\"\\"\\"another description\\"\\"\\"
  ANOTHER
}

\\"\\"\\"This is a test description\\"\\"\\"
type TestCollection {
  totalCount: Int!
  nodes: Int
}

type TestObjectWithinAList {
  testListOfObjects: ID!
  testModel: User
  enum: SomeTestEnum
}

\\"\\"\\"Some user here\\"\\"\\"
type User {
  id: String
  name: String
}
"
`;
